### PR TITLE
Use consistent latest block number across whole run, use block numbers instead of block hashes

### LINKED ISSUE #1049: Use consistent latest block number across whole test suite
I think this is the kind of information we should have before even running the test, so:
1. It's consistent all across the board (whole suite)
2. We don't have to return it from the test itself

Can we please address this?

_Originally posted by @Arcticae in https://github.com/foundry-rs/starknet-foundry/pull/1007#discussion_r1383032262_

### PR DESCRIPTION
**Pull request by piotmag769**
## Introduced changes

- Use consistent latest block number for each url across whole run - closes #1049 
- Use block numbers instead of hashes to enable reusing cache (same block can be referenced by both hash and number, but the cache from number cannot be reused for hash and the other way round) - partially solves #1102 (will open another PR to enforce proper types etc.)
- Latest block numbers for each url are now printed at the end (if `BlockId::Tag(Latest)` was specified for such url)

## Checklist

- [x] Linked relevant issue
- [x] Updated relevant documentation
- [x] Added relevant tests
- [x] Performed self-review of the code
- [x] Added changes to `CHANGELOG.md`

### CODE REVIEW

File: crates/cheatnet/tests/contracts/Scarb.lock [hunk 1]
```diff
@@ -1,6 +0,0 @@
-# Code generated by scarb DO NOT EDIT.
-version = 1
-
-[[package]]
-name = "cheatnet_testing_contracts"
-version = "0.1.0"
```
Comment (piotmag769) [inline crates/cheatnet/tests/contracts/Scarb.lock:6]: It got removed when clearing git cache and readding everything, shouldn't have been there in the first place

File: crates/forge/src/block_number_map.rs [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+    pub fn get_url_to_latest_block_number(&self) -> &HashMap<String, BlockNumber> {
+        &self.url_to_latest_block_number
+    }
+}
+
+async fn get_latest_block_number(url: &Url) -> Result<BlockNumber> {
+    let client = JsonRpcClient::new(HttpTransport::new(url.clone()));
+
+    match Handle::current()
+        .spawn(async move { client.get_block_with_tx_hashes(BlockId::Tag(Latest)).await })
```
Comment (piotmag769) [inline crates/forge/src/block_number_map.rs:60]: Btw this is the solution that should be used to solve https://github.com/foundry-rs/starknet-foundry/issues/901
cc @cptartur @tkumor3 @Arcticae @MaksymilianDemitraszek

File: crates/forge/src/block_number_map.rs [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+            assert_eq!(fork_params_string.block_id_value, "Latest");
+            if let Some(block_number) = block_number_map.get_latest_block_number(&url_str) {
+                *block_number
+            } else {
+                let latest_block_number = get_latest_block_number(&url).await?;
+                block_number_map.add_latest_block_number(url_str, latest_block_number);
+                latest_block_number
+            }
+        }
+        _ => unreachable!(),
```
Comment (Arcticae) [inline crates/forge/src/block_number_map.rs:115]: Can we program this so it does not use `unreachable!` macro?

File: crates/forge/src/block_number_map.rs [hunk 4]
```diff
@@ -0,0 +1,118 @@
...
+        }
+        "tag" => {
+            assert_eq!(fork_params_string.block_id_value, "Latest");
+            if let Some(block_number) = block_number_map.get_latest_block_number(&url_str) {
+                *block_number
+            } else {
+                let latest_block_number = get_latest_block_number(&url).await?;
+                block_number_map.add_latest_block_number(url_str, latest_block_number);
+                latest_block_number
+            }
```
Comment (Arcticae) [inline crates/forge/src/block_number_map.rs:113]: Shouldn't fetching this be responsibility of the class that caches the results? I find it confusing to have a standalone function do this

File: crates/forge/src/block_number_map.rs (hunk 3 above)
Comment (piotmag769) [reply crates/forge/src/block_number_map.rs:115]: I can throw a panic here or add an error message, but idk why it changes anything. Those are verified in scarb test collector or when parsing config from Scarb.toml

File: crates/forge/src/block_number_map.rs (hunk 4 above)
Comment (piotmag769) [reply crates/forge/src/block_number_map.rs:113]: Can do that

**Review by Arcticae — APPROVED** [review]

**Pull request merged**